In [12]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from collections import Counter

df = pd.read_csv("/Users/margauxelbaum/Desktop/BossLadies/archive/cleaned_ingredients.csv")

df.head()



,NDB_No,Descrip,Energy_kcal,Protein_g,Saturated_fats_g,Fat_g,Carb_g,Fiber_g,Sugar_g,Calcium_mg,Iron_mg,Magnesium_mg,Phosphorus_mg,Potassium_mg,Sodium_mg,Zinc_mg,Copper_mcg,Manganese_mg,Selenium_mcg,VitC_mg,Thiamin_mg,Riboflavin_mg,Niacin_mg,VitB6_mg,Folate_mcg,VitB12_mcg,VitA_mcg,VitE_mg,VitD2_mcg
0,1001,butter with salt,717.0,0.85,51.368,81.11,0.06,0.0,0.06,24.0,0.02,2,24.0,24.0,643.0,0.09,0.000,0.000,1.0,0.0,0.005,0.034,0.042,0.003,3.0,0.17,684.0,2.32,0.0
1,1002,butter whipped w salt,718.0,0.49,45.390,78.30,2.87,0.0,0.06,23.0,0.05,1,24.0,41.0,583.0,0.05,0.010,0.001,0.0,0.0,0.007,0.064,0.022,0.008,4.0,0.07,683.0,1.37,0.0
2,1003,butter oil anhydrous,876.0,0.28,61.924,99.48,0.00,0.0,0.00,4.0,0.00,0,3.0,5.0,2.0,0.01,0.001,0.000,0.0,0.0,0.001,0.005,0.003,0.001,0.0,0.01,840.0,2.8,0.0
3,1004,cheese blue,353.0,21.40,18.669,28.74,2.34,0.0,0.50,528.0,0.31,23,387.0,256.0,1146.0,2.66,0.040,0.009,14.5,0.0,0.029,0.382,1.016,0.166,36.0,1.22,198.0,0.25,0.5
4,1005,cheese brick,371.0,23.24,18.764,29.68,2.79,0.0,0.51,674.0,0.43,24,451.0,136.0,560.0,2.60,0.024,0.012,14.5,0.0,0.014,0.351,0.118,0.065,20.0,1.26,292.0,0.26,0.5


In [3]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from collections import Counter

df = pd.read_csv("/Users/margauxelbaum/Desktop/BossLadies/archive/cleaned_ingredients.csv")

# ── 1. NaN values ──────────────────────────────────────────────────────────────
print("=" * 60)
print("1. NaN VALUES")
print("=" * 60)
nan_counts = df.isnull().sum()
nan_pct = (df.isnull().mean() * 100).round(2)
nan_df = pd.DataFrame({"NaN count": nan_counts, "NaN %": nan_pct})
nan_df = nan_df[nan_df["NaN count"] > 0].sort_values("NaN count", ascending=False)
if nan_df.empty:
    print("No NaN values found!")
else:
    print(nan_df.to_string())
print(f"\nTotal NaN cells: {df.isnull().sum().sum()}")

# ── 2. Min / Max per numeric column ────────────────────────────────────────────
print("\n" + "=" * 60)
print("2. MIN / MAX PER COLUMN")
print("=" * 60)
numeric_cols = df.select_dtypes(include=[np.number]).columns
minmax = pd.DataFrame({
    "min": df[numeric_cols].min(),
    "max": df[numeric_cols].max()
})
print(minmax.to_string())

# ── 3. Non-feasible values ─────────────────────────────────────────────────────
print("\n" + "=" * 60)
print("3. NON-FEASIBLE VALUES")
print("=" * 60)

# 3a. Energy > 900 kcal
if "Energy_kcal" in df.columns:
    mask_kcal = df["Energy_kcal"] > 900
    print(f"\n[Energy > 900 kcal]  {mask_kcal.sum()} rows")
    if mask_kcal.sum():
        cols_show = ["NDB_No", "Descrip", "Energy_kcal"] if "Descrip" in df.columns else ["NDB_No", "Energy_kcal"]
        print(df.loc[mask_kcal, [c for c in cols_show if c in df.columns]].to_string(index=False))

# 3b. g-per-100g columns that exceed 100
over100_cols = ["Protein_g", "Fat_g", "Saturated_fats_g", "Carb_g", "Fiber_g", "Sugar_g"]
over100_cols = [c for c in over100_cols if c in df.columns]

for col in over100_cols:
    mask = df[col] > 100
    print(f"\n[{col} > 100 g]  {mask.sum()} rows")
    if mask.sum():
        id_cols = [c for c in ["NDB_No", "Descrip"] if c in df.columns]
        print(df.loc[mask, id_cols + [col]].to_string(index=False))

# 3c. Any negative values in numeric columns
print("\n[Negative values in numeric columns]")
neg_found = False
for col in numeric_cols:
    neg_mask = df[col] < 0
    if neg_mask.sum() > 0:
        neg_found = True
        print(f"  {col}: {neg_mask.sum()} negative rows")
if not neg_found:
    print("  None found.")

print("\n" + "=" * 60)
print("Done.")

1. NaN VALUES
No NaN values found!

Total NaN cells: 0

2. MIN / MAX PER COLUMN
                  min        max
Energy_kcal       0.0    902.000
Protein_g         0.0     88.320
Saturated_fats_g  0.0     95.600
Fat_g             0.0    100.000
Carb_g            0.0    100.000
Fiber_g           0.0     79.000
Sugar_g           0.0     99.800
Calcium_mg        0.0   7364.000
Iron_mg           0.0    123.600
Phosphorus_mg     0.0   9918.000
Potassium_mg      0.0  16500.000
Sodium_mg         0.0  38758.000
Zinc_mg           0.0     90.950
Copper_mcg        0.0     15.050
Manganese_mg      0.0    328.000
Selenium_mcg      0.0   1917.000
VitC_mg           0.0   2732.000
Thiamin_mg        0.0     23.375
Riboflavin_mg     0.0     17.500
Niacin_mg         0.0    127.500
VitB6_mg          0.0    316.000
Folate_mcg        0.0   2875.000
VitB12_mcg        0.0     98.890
VitA_mcg          0.0  30000.000
VitD2_mcg         0.0    250.000

3. NON-FEASIBLE VALUES

[Energy > 900 kcal]  9 rows
NDB_No   

We can see that all are capped at max 100g for the important columns, now I want to see whether the sum all macros doesn't exceed 100g (knowing that both fiber and sugar are a subset of carbs)

In [10]:
# Correct macro sum: no double-counting
g_cols_correct = ["Protein_g", "Fat_g", "Carb_g"]
df["macro_sum_correct"] = df[g_cols_correct].sum(axis=1)
mask_correct = df["macro_sum_correct"] > 100

print(f"[Protein + Fat + Carb only] Rows > 100g: {mask_correct.sum()}")

if mask_correct.sum() > 0:
    id_cols = [c for c in ["NDB_No", "Descrip"] if c in df.columns]
    print(df.loc[mask_correct, id_cols + g_cols_correct + ["macro_sum_correct"]].to_string(index=False))

[Protein + Fat + Carb only] Rows > 100g: 2
NDB_No                                   Descrip  Protein_g  Fat_g  Carb_g  macro_sum_correct
 35057 oil bearded seal (oogruk) (alaska native)       0.56  99.60     0.0             100.16
 42231                 oil flaxseed cold pressed       0.11  99.98     0.0             100.09


We can see that there are only 2 rows that are not exactly capped at 100, and those are only a little above 100g. Most the values are under 100g, but this is probably due to the water in them

Next thing we need to know, is how many food items are at 0.0 for one of the macros, this could either mean those values are truly zeros, or they are NaN, and we need to handle them 

In [16]:
numeric_cols = df.select_dtypes(include=[np.number]).columns.tolist()
if "NDB_No" in numeric_cols:
    numeric_cols.remove("NDB_No")
numeric_cols = pd.Index(numeric_cols)

print(numeric_cols)  # verify before proceeding

zero_counts = (df[numeric_cols] == 0.0).sum()
zero_pct = ((df[numeric_cols] == 0.0).mean() * 100).round(2)

zero_df = pd.DataFrame({
    "zero count": zero_counts,
    "zero %": zero_pct,
    "NaN count": df[numeric_cols].isnull().sum(),
    "NaN %": (df[numeric_cols].isnull().mean() * 100).round(2)
}).sort_values("zero %", ascending=False)

print("Zeros vs NaNs per column:")
print(zero_df.to_string())

Index(['Energy_kcal', 'Protein_g', 'Saturated_fats_g', 'Fat_g', 'Carb_g',
       'Fiber_g', 'Sugar_g', 'Calcium_mg', 'Iron_mg', 'Phosphorus_mg',
       'Potassium_mg', 'Sodium_mg', 'Zinc_mg', 'Copper_mcg', 'Manganese_mg',
       'Selenium_mcg', 'VitC_mg', 'Thiamin_mg', 'Riboflavin_mg', 'Niacin_mg',
       'VitB6_mg', 'Folate_mcg', 'VitB12_mcg', 'VitA_mcg', 'VitD2_mcg'],
      dtype='object')
Zeros vs NaNs per column:
                  zero count  zero %  NaN count  NaN %
VitD2_mcg               6998   75.10          0    0.0
VitC_mg                 5209   55.90          0    0.0
VitB12_mcg              4657   49.98          0    0.0
VitA_mcg                4355   46.74          0    0.0
Fiber_g                 4275   45.88          0    0.0
Sugar_g                 4180   44.86          0    0.0
Folate_mcg              2568   27.56          0    0.0
Carb_g                  2372   25.46          0    0.0
Manganese_mg            2315   24.84          0    0.0
Selenium_mcg            2178 

In [18]:
check_cols = ["Energy_kcal", "Protein_g", "Fat_g", "Carb_g", "Fiber_g", "Sugar_g"]

for col in check_cols:
    mask_zero = df[col] == 0.0
    n_zeros = mask_zero.sum()
    print(f"\n{'='*60}")
    print(f"{col} — {n_zeros} zeros ({n_zeros/len(df)*100:.1f}%)")
    if n_zeros > 0:
        print(df.loc[mask_zero, ["Descrip"] + check_cols].to_string(index=False))

# ── Rows where ALL three main macros are 0 ─────────────────────────────────────
print(f"\n{'='*60}")
mask_all_zero = (df["Protein_g"] == 0.0) & (df["Fat_g"] == 0.0) & (df["Carb_g"] == 0.0)
print(f"Rows where Protein + Fat + Carb are ALL 0.0: {mask_all_zero.sum()}")
if mask_all_zero.sum() > 0:
    print(df.loc[mask_all_zero, ["Descrip"] + check_cols].to_string(index=False))


Energy_kcal — 39 zeros (0.4%)
                                                     Descrip  Energy_kcal  Protein_g  Fat_g  Carb_g  Fiber_g  Sugar_g
                                                  salt table          0.0        0.0    0.0    0.00      0.0     0.00
                  seasoning mix dry sazon coriander  annatto          0.0        0.0    0.0    0.00      0.0     0.00
                  babyfood h2o btld gerber wo added fluoride          0.0        0.0    0.0    0.00      0.0     0.00
                                        beverages zevia cola          0.0        0.0    0.0    1.13      0.0     0.00
                          beverages zevia cola caffeine free          0.0        0.0    0.0    1.13      0.0     0.00
  beverage gmbh  co kg gerolsteiner naturally spark min h2o           0.0        0.0    0.0    0.00      0.0     0.00
                  beverages icelandic glacial nat spring h2o          0.0        0.0    0.0    0.00      0.0     0.00
   bev gerolsteiner brunn

I also wanted to see, which food items had their sum at 0.0, to see if those were missing values or true. 
All of those are either pure minerals, water (different brands), zero sports drinks, zero calories beverages or seasonix mix. Also, those 39 rows correspond to the number of rows that kcal also has as 0.  

In [21]:
check_cols = ["Energy_kcal", "Protein_g", "Fat_g", "Carb_g", "Fiber_g", "Sugar_g"]

# collect all rows with at least one zero in macro cols
mask_any_zero = (df[check_cols] == 0.0).any(axis=1)
df_zero_items = df.loc[mask_any_zero, ["NDB_No", "Descrip"] + check_cols].copy()

# add a column showing WHICH macros are zero for each row
df_zero_items["zero_in"] = df.loc[mask_any_zero, check_cols].apply(
    lambda row: ", ".join([col for col in check_cols if row[col] == 0.0]), axis=1
)

print(f"Total rows with at least one macro at 0: {len(df_zero_items)}")

# save to csv
df_zero_items.to_csv("/Users/margauxelbaum/Desktop/BossLadies/zero_macro_items.csv", index=False)
print("Saved to /Users/margauxelbaum/Desktop/BossLadies/zero_macro_items.csv")

Total rows with at least one macro at 0: 5160
Saved to /Users/margauxelbaum/Desktop/BossLadies/zero_macro_items.csv


After checking the csv, I think most of the values are correct, particularly for fiber, as most of the missing values in the csv file are fiber, but lots of items don't contain any fiber. However, one column that needs a bit more attention is sugar. As some items are highly probable to have sugar in them, such as muffin, donuts, pie, etc. However, most of the values for sugar are still correct, as many columns include different types of meat. 

In [26]:
suspicious_keywords = ["muffin", "donut", "doughnut", "pie", "cake", "cookie", 
                       "brownie", "pastry", "waffle", "pancake", "syrup", "candy",
                       "chocolate", "ice cream", "pudding", "tart", "croissant", "granola bar"]

pattern = "|".join(suspicious_keywords)

mask_suspicious_sugar = (
    (df["Sugar_g"] == 0.0) & 
    (df["Descrip"].str.contains(pattern, case=False, na=False))
)

print(f"Suspicious sugar zeros: {mask_suspicious_sugar.sum()}")
print(df.loc[mask_suspicious_sugar, ["Descrip", "Energy_kcal", "Carb_g", "Sugar_g"]].to_string(index=False))

df["sugar_unreliable"] = mask_suspicious_sugar

Suspicious sugar zeros: 181
                                                     Descrip  Energy_kcal  Carb_g  Sugar_g
                                  babyfood cookies arrowroot        424.0   75.40      0.0
               shortening cake mix soybn (hydr)cttnsd (hydr)        884.0    0.00      0.0
      shortening spl purpose for cakesfrostings soybn (hydr)        884.0    0.00      0.0
  margarine indus soy  part hydr soy oil baking sauces candy        714.0    0.71      0.0
                 apricots cnd ex lt syrup pk wskn solliquids         49.0   12.50      0.0
                  apricots cnd hvy syrup pk woskn solliquids         83.0   21.45      0.0
               apricots cnd ex hvy syrup pk woskn solliquids         96.0   24.85      0.0
                                 boysenberries cnd hvy syrup         88.0   22.31      0.0
                cherries sour red cnd lt syrup pk solliquids         75.0   19.30      0.0
            cherries sour red cnd ex hvy syrup pk solliquids  

Some of those values are definitely suspicious, such as syrup, which is mainly sugar, however "ex lt" means extra light, but they should still contain some sugar. 
One option would be to flag those, don't rely on them, except if there is a "sugar free" indication?

In [27]:
suspicious_keywords = [
    "muffin", "donut", "doughnut", "pie", "cake", "cookie",
    "brownie", "pastry", "waffle", "pancake", "candy",
    "chocolate", "ice cream", "pudding", "tart", "croissant",
    "syrup", "agutuk", "arrowroot babyfood"
]

pattern = "|".join(suspicious_keywords)

mask_suspicious_sugar = (
    (df["Sugar_g"] == 0.0) &
    (df["Descrip"].str.contains(pattern, case=False, na=False)) &
    (~df["Descrip"].str.contains("sugar free|sugarfree|no sugar|unsweetened", case=False, na=False))
)

df["sugar_unreliable"] = mask_suspicious_sugar

print(f"Flagged as sugar_unreliable: {mask_suspicious_sugar.sum()} rows")
print(df.loc[mask_suspicious_sugar, ["Descrip", "Carb_g", "Sugar_g"]].to_string(index=False))

Flagged as sugar_unreliable: 167 rows
                                                     Descrip  Carb_g  Sugar_g
                                  babyfood cookies arrowroot   75.40      0.0
               shortening cake mix soybn (hydr)cttnsd (hydr)    0.00      0.0
      shortening spl purpose for cakesfrostings soybn (hydr)    0.00      0.0
  margarine indus soy  part hydr soy oil baking sauces candy    0.71      0.0
                 apricots cnd ex lt syrup pk wskn solliquids   12.50      0.0
                  apricots cnd hvy syrup pk woskn solliquids   21.45      0.0
               apricots cnd ex hvy syrup pk woskn solliquids   24.85      0.0
                                 boysenberries cnd hvy syrup   22.31      0.0
                cherries sour red cnd lt syrup pk solliquids   19.30      0.0
            cherries sour red cnd ex hvy syrup pk solliquids   29.23      0.0
                 cherries swt cnd ex hvy syrup pk solliquids   26.23      0.0
                         f

Out of the 181 values, 14 are unsweetened, so truly 0 sugar. 

Overall, this is a very good dataset, much more reliable than the other one. 